### PR TITLE
[RISCV] Add 32 bit GPR sub-register for Zfinx.

### PR DESCRIPTION
**Pull request by topperc**
This patches adds a 32 bit register class for use with Zfinx instructions.
This makes them more similar to F instructions and allows us to
only spill 32 bits.
    
I've added CodeGenOnly instructions for load/store using GPRF32 as that
gave better results than insert_subreg/extract_subreg. I'm using FSGNJ for
GPRF32 copy with Zfinx as that gave better results from MachineCopyPropagation.
    
Function arguments use this new GPRF32 register class for f32 arguments
with Zfinx. Eliminating the need to use RISCVISD::FMV* nodes.
    
This is similar to #107446 which adds a 16 bit register class.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVCallingConv.cpp
Comment (dtcxzyw) [inline llvm/lib/Target/RISCV/RISCVCallingConv.cpp:787]:
```suggestion
  if ((LocVT == MVT::f64 && Subtarget.hasStdExtZdinx() &&
```

File: llvm/lib/Target/RISCV/RISCVInstrInfoF.td
Comment (dtcxzyw) [inline llvm/lib/Target/RISCV/RISCVInstrInfoF.td:313]:
```suggestion
// ADDI with GPRF32 register class to use for copy. This should not be used as
```

File: llvm/lib/Target/RISCV/RISCVCallingConv.cpp
Comment (dtcxzyw) [inline llvm/lib/Target/RISCV/RISCVCallingConv.cpp:404]:
```suggestion
  if (ValVT == MVT::f32 && Subtarget.hasStdExtZfinx()) {
```

File: llvm/lib/Target/RISCV/RISCVCallingConv.cpp
Comment (dtcxzyw) [inline llvm/lib/Target/RISCV/RISCVCallingConv.cpp:787]:
```suggestion
  if (LocVT == MVT::f64 && Subtarget.hasStdExtZdinx() &&
       Subtarget.is64Bit()) {
```

**Review by dtcxzyw — APPROVED** [review]
LGTM. Please wait for additional approval from other reviewers :)

File: llvm/lib/Target/RISCV/RISCVCallingConv.cpp
Comment (asb) [inline llvm/lib/Target/RISCV/RISCVCallingConv.cpp:165]: ILP64E => LP64E

File: llvm/lib/Target/RISCV/RISCVCallingConv.cpp
Comment (asb) [inline llvm/lib/Target/RISCV/RISCVCallingConv.cpp:775]: Not really an issue for this review directly, but I wonder if we should be more aggressive about erroring out of the GHC calling convention for more obscure targets that are unlikely to ever need it. Is the ghc cc realistically ever going to be needed on a zfinx/zdinx system?

Change looks fine though.

**Review by asb — APPROVED** [review]
LGTM.

**Pull request merged**
